File: supabase/functions/_shared/macroEngine.ts
// Macro calculation shared by strava-webhook and runna-sync.
// Mirrors CRUNCH/Engines/MacroEngine.swift — master spec §3–6 (Section 14 items 1–6).
// Scope: targeted single-value engine upgrade. Training-level carb bands (§4.1
// columns), race/age/diet modifiers, phase carb multipliers, and the §7.1 phase
// detection (post_race_recovery) are deferred to the Section 7–12 phase.

import type { SupabaseClient } from "https://esm.sh/@supabase/supabase-js@2";

// ── Constants ─────────────────────────────────────────────────────────────────

// TDEE multipliers (§3.2). Existing run values unchanged; recovery_day + race_*
// split are the in-scope additions.
const TDEE_MULTIPLIER: Record<string, number> = {
  rest:          1.2,
  recovery_day:  1.35,
  easy_run:      1.55,
  tempo:         1.725,
  interval:      1.725,
  long_run:      1.9,
  race_5k:       1.75,
  race_10k:      1.85,
  race_half:     2.00,
  race_marathon: 2.40,
};
const DEFAULT_MULTIPLIER = 1.2; // rest

// Carbohydrate g/kg (§4.1, Band-B representative).
const CARBS_PER_KG: Record<string, number> = {
  rest:          4,
  recovery_day:  6,
  easy_run:      6,
  tempo:         7,
  interval:      7,
  long_run:      8.5,
  race_5k:       5.5,
  race_10k:      6.5,
  race_half:     8.5,
  race_marathon: 10,
};

// Session carb floors (§6.3) — lower bound both fat guards reduce toward.
const SESSION_CARB_FLOOR: Record<string, number> = {
  long_run:      5.0,
  race_half:     6.0,
  race_marathon: 8.0,
  race_ultra:    8.0,
};
const DEFAULT_CARB_FLOOR = 3.0;

const PROTEIN_PER_KG          = 1.7;   // Morton et al. 2018
const RECOVERY_PROTEIN_PER_KG = 2.0;   // Ivy 2002 (recovery-day only; §5.1 subset)
const RACE_WEEK_CARB_LOAD     = 11;    // flat carb-load; §4.5 race-specific values deferred
const KCAL_CARBS              = 4;
const KCAL_PROTEIN            = 4;
const KCAL_FAT                = 9;
const FAT_MIN_PCT             = 0.20;  // 20% of energy (not 0.5 g/kg — §6.2)
const FAT_MAX_PCT             = 0.35;

const FALLBACK = {
  weight_kg: 70,
  height_cm: 175,   // matches Swift fallback (was 170)
  age:       30,
  gender:    "male",
};

const SESSION_TO_TARGET_TYPE: Record<string, string> = {
  rest:          "rest",
  recovery_day:  "recovery",
  easy_run:      "easy",
  tempo:         "tempo",
  interval:      "interval",
  long_run:      "long",
  race_5k:       "race_5k",
  race_10k:      "race_10k",
  race_half:     "race_half",
  race_marathon: "race_marathon",
};

const HARD_SESSION_TYPES = new Set([
  "tempo", "interval", "long_run",
  "race", "race_5k", "race_10k", "race_half", "race_marathon", "race_ultra",
]);
const RUN_SESSION_TYPES = new Set([
  "easy_run", "tempo", "interval", "long_run",
  "race", "race_5k", "race_10k", "race_half", "race_marathon", "race_ultra",
]);

// ── Session-type resolution ─────────────────────────────────────────────────────

// Maps aliases onto canonical lookup keys. Legacy/ultra "race" folds to
// race_marathon so it can never fall through to the rest default (defensive alias).
function canonical(type: string): string {
  if (type === "cycling" || type === "swimming") return "easy_run";
  if (type === "race" || type === "race_ultra" || type === "race_ultra_marathon") {
    return "race_marathon";
  }
  return type;
}

function isRaceType(type: string): boolean {
  return type === "race" || type.startsWith("race_");
}

// Recovery-day detection (§3.4). Upgrades a light day that follows a long run or
// race; deliberately does not downgrade a genuine hard session (would under-fuel).
function resolveSessionType(today: string, previous: string | null): string {
  if (previous && (previous === "long_run" || previous === "race" || previous.startsWith("race_"))) {
    return HARD_SESSION_TYPES.has(today) ? today : "recovery_day";
  }
  return today;
}

// ── Pure calculation ──────────────────────────────────────────────────────────

type TrainingPhase =
  | "base_building"
  | "build_phase"
  | "peak_training"
  | "taper"
  | "race_week";

interface MacroResult {
  calories_kcal:  number;
  carbs_g:        number;
  protein_g:      number;
  fat_g:          number;
  target_type:    string;
  training_phase: TrainingPhase;
  flags:          string[];
}

function getDaysToRace(raceDate: string, sessionDate: string): number {
  const race    = new Date(raceDate);
  const session = new Date(sessionDate);
  race.setUTCHours(0, 0, 0, 0);
  session.setUTCHours(0, 0, 0, 0);
  return Math.round((race.getTime() - session.getTime()) / 86_400_000);
}

// Matches Swift MacroEngine.trainingPhase: >12 base, 8–12 build, 4–7 peak,
// 1–3 taper, 0 race week. Weeks are floored and clamped at 0 (Swift weeksUntil).
function getPhase(daysToRace: number): TrainingPhase {
  const weeks = Math.max(0, Math.floor(daysToRace / 7));
  if (weeks > 12) return "base_building";
  if (weeks >= 8) return "build_phase";
  if (weeks >= 4) return "peak_training";
  if (weeks >= 1) return "taper";
  return "race_week";
}

function calcBmr(weight: number, height: number, age: number, gender: string): number {
  const base = 10 * weight + 6.25 * height - 5 * age;
  return gender === "female" ? base - 161 : base + 5;
}

// Fat Engine (§6.1): 20–35%-of-energy band + FIX A (ordinary fat-floor
// reconciliation) + FIX B (carb-load collision guard).
function calcFat(
  tdee: number, carbs: number, protein: number, weight: number, type: string,
): { fat: number; carbs: number; flags: string[] } {
  const usedCalories = carbs * KCAL_CARBS + protein * KCAL_PROTEIN;
  const fatFromRemainder = Math.max(0, (tdee - usedCalories) / KCAL_FAT);

  const fatMin = (tdee * FAT_MIN_PCT) / KCAL_FAT;
  const fatMax = (tdee * FAT_MAX_PCT) / KCAL_FAT;

  let fat = Math.max(fatMin, Math.min(fatFromRemainder, fatMax));
  let finalCarbs = carbs;
  const flags: string[] = [];

  // FIX A — reduce carbs toward the session floor when the fat floor triggers,
  // then re-derive fat from the room actually left. Every session type.
  if (fat > fatFromRemainder) {
    flags.push("fat_floor_triggered");
    const carbReduction = ((fat - fatFromRemainder) * KCAL_FAT) / KCAL_CARBS;
    const sessionFloor = (SESSION_CARB_FLOOR[type] ?? DEFAULT_CARB_FLOOR) * weight;
    finalCarbs = Math.max(carbs - carbReduction, sessionFloor);
    const remainingForFat = tdee - (finalCarbs * KCAL_CARBS + protein * KCAL_PROTEIN);
    fat = Math.max(remainingForFat / KCAL_FAT, fatMin * 0.9);
  }

  // FIX B — carb-load collision guard, TDEE-relative floor (not flat 8 g/kg).
  const totalCheck = finalCarbs * KCAL_CARBS + protein * KCAL_PROTEIN + fatMin * KCAL_FAT;
  if (totalCheck > tdee * 1.02) {
    const reduction = (totalCheck - tdee) / KCAL_CARBS;
    const hardFloor = Math.min(8.0 * weight, (tdee * 0.55) / KCAL_CARBS);
    const newCarbs = Math.max(finalCarbs - reduction, hardFloor);
    if (Math.abs(newCarbs - finalCarbs) > 1e-6) {
      flags.push("carb_load_capped");
      finalCarbs = newCarbs;
      const remainingForFat = tdee - (finalCarbs * KCAL_CARBS + protein * KCAL_PROTEIN);
      fat = Math.max(remainingForFat / KCAL_FAT, fatMin * 0.9);
    }
  }

  return { fat, carbs: finalCarbs, flags };
}

function calcMacros(
  weight: number, height: number, age: number, gender: string,
  sessionType: string, previousSessionType: string | null,
  sessionDate: string, raceDate: string,
): MacroResult {
  const resolved = resolveSessionType(sessionType, previousSessionType);
  const type = canonical(resolved);

  const tdee = calcBmr(weight, height, age, gender) * (TDEE_MULTIPLIER[type] ?? DEFAULT_MULTIPLIER);
  const phase = getPhase(getDaysToRace(raceDate, sessionDate));

  // Race week forces the flat carb-load; otherwise the session's g/kg.
  const baseCarbs = phase === "race_week"
    ? RACE_WEEK_CARB_LOAD * weight
    : (CARBS_PER_KG[type] ?? CARBS_PER_KG["easy_run"]) * weight;
  const protein = (type === "recovery_day" ? RECOVERY_PROTEIN_PER_KG : PROTEIN_PER_KG) * weight;

  const { fat, carbs, flags } = calcFat(tdee, baseCarbs, protein, weight, type);

  const targetType = phase === "race_week"
    ? "carb_load"
    : (SESSION_TO_TARGET_TYPE[resolved] ?? "easy");

  return {
    calories_kcal:  Math.round(carbs * KCAL_CARBS + protein * KCAL_PROTEIN + fat * KCAL_FAT),
    carbs_g:        Math.round(carbs),
    protein_g:      Math.round(protein),
    fat_g:          Math.round(fat),
    target_type:    targetType,
    training_phase: phase,
    flags,
  };
}

// ── DB helper ─────────────────────────────────────────────────────────────────

function previousDate(sessionDate: string): string {
  const d = new Date(sessionDate);
  d.setUTCHours(0, 0, 0, 0);
  d.setUTCDate(d.getUTCDate() - 1);
  return d.toISOString().substring(0, 10);
}

export async function generateAndSaveMacroTarget(
  supabase: SupabaseClient,
  params: {
    userId:      string;
    sessionId:   string;
    sessionDate: string;
    sessionType: string;   // already race-resolved by the caller (see strava-webhook)
  },
): Promise<void> {
  const { userId, sessionId, sessionDate, sessionType } = params;

  const [userRes, raceRes, prevRes] = await Promise.all([
    supabase
      .from("users")
      .select("weight_kg, height_cm, age, gender")
      .eq("id", userId)
      .single(),
    supabase
      .from("races")
      .select("race_date")
      .eq("user_id", userId)
      .eq("is_active", true)
      .single(),
    // Yesterday's run — drives recovery-day detection (§3.4).
    supabase
      .from("training_sessions")
      .select("session_type")
      .eq("user_id", userId)
      .eq("session_date", previousDate(sessionDate)),
  ]);

  if (userRes.error || !userRes.data || raceRes.error || !raceRes.data) return;

  const u = userRes.data as {
    weight_kg: number | null; height_cm: number | null;
    age: number | null; gender: string | null;
  };
  const raceDate = (raceRes.data as { race_date: string }).race_date;

  const prevRows = (prevRes.data ?? []) as { session_type: string }[];
  const previousSessionType =
    prevRows.find((r) => RUN_SESSION_TYPES.has(r.session_type))?.session_type ?? null;

  const result = calcMacros(
    u.weight_kg ?? FALLBACK.weight_kg,
    u.height_cm ?? FALLBACK.height_cm,
    u.age       ?? FALLBACK.age,
    u.gender    ?? FALLBACK.gender,
    sessionType,
    previousSessionType,
    sessionDate,
    raceDate,
  );

  await supabase.from("macro_targets").upsert(
    {
      user_id:       userId,
      target_date:   sessionDate,
      session_id:    sessionId,
      calories_kcal: result.calories_kcal,
      carbs_g:       result.carbs_g,
      protein_g:     result.protein_g,
      fat_g:         result.fat_g,
      target_type:   result.target_type,
    },
    { onConflict: "user_id,target_date" },
  );
}

// Distance-bucket race resolution (master spec flag 2, distance PRIMARY).
// Callers pass a bare "race" (e.g. Strava workout_type=1); bucket by the actual
// session distance so a marathon-training user's 10K tune-up gets race_10k fuel.
// distance_km null → race_marathon (the safe default; the race_type fallback is
// applied by the legacy-row backfill migration, where distance may be absent).
export function resolveRaceSessionType(sessionType: string, distanceKm: number | null): string {
  if (!isRaceType(sessionType)) return sessionType;
  if (distanceKm == null) return "race_marathon";
  if (distanceKm < 7)  return "race_5k";
  if (distanceKm < 15) return "race_10k";
  if (distanceKm < 25) return "race_half";
  return "race_marathon";
}

File: CRUNCH/Engines/MacroEngine.swift
import Foundation

enum TrainingPhase: String {
    case baseBuilding = "Base Building"
    case build        = "Build"
    case peakTraining = "Peak Training"
    case taper        = "Taper"
    case raceWeek     = "Race Week"
}

// Pure calculation engine — no state, no side effects.
// All formulas per docs/crunch-nutrition-engine-MASTER-SPEC.md §3–6:
// BMR: Mifflin-St Jeor (Mifflin 1990)
// Carb targets: Burke et al. 2011, Jeukendrup 2011
// Protein: Morton et al. BJSM 2018, ISSN 2017; recovery-day 2.0 g/kg (Ivy 2002)
// Fat: 20–35% of energy with reconciliation guards (Mountjoy 2014, Loucks 2004)
//
// Scope note (master-spec Sections 14 items 1–7 + the §14.11 "confirm wired"
// tail): targeted upgrade of the single-value engine, NOT the full 8-layer
// rebuild. Now switched on (Phase 5.1): training-level A/B/C carb bands (§4.1),
// day-before long-run/race carb boost (§4.2, surfaced via Layer 8), and the
// masters age protein modifier with the §5.1 2.5 g/kg hard cap. Already live:
// diet protein modifier (§2.4) + low-carb conflict flag (§9.2). Still deferred
// to the Section 7–12 phase: phase carb/protein multipliers (§4.3), the new
// phase detection (§7.1 post_race_recovery), and the duration/MET model (§3.3).
enum MacroEngine {

    // MARK: - Public

    static func calculate(
        user: UserProfile,
        raceDate: String?,
        sessionType: String,
        previousSessionType: String? = nil,
        nextSessionType: String? = nil,
        additionalActivities: [ActivityType] = []
    ) -> MacroTarget {
        let kg = user.weightKg

        // Training-level band (§2.1) — selects the Layer 3 carb-table column.
        let band = trainingBand(user.trainingLevel)

        // Resolve recovery-day (§3.4), then canonicalise aliases (cycling/swimming
        // → easy_run; legacy/ultra "race" → race_marathon).
        let resolvedType = resolveSessionType(today: sessionType, previous: previousSessionType)
        let type = canonical(resolvedType)

        // BMR (Mifflin-St Jeor)
        let bmr: Double = user.gender == "female"
            ? 10 * kg + 6.25 * user.heightCm - 5 * Double(user.age) - 161
            : 10 * kg + 6.25 * user.heightCm - 5 * Double(user.age) + 5

        // TDEE
        let tdee = bmr * tdeeMultiplier(type)

        // Training phase.
        // TODO(Section 7 — Phase Engine, deferred): base/build/peak/taper
        // currently compute identically for a given session_type; only
        // race_week diverges (via the carb-load protocol). This is correct
        // interim behavior, not a bug — full phase-based carb/protein
        // multipliers land when Sections 7.1-7.2 are implemented.
        let weeks = raceDate.map { weeksUntil(dateString: $0) } ?? 20
        let phase = trainingPhase(weeksToRace: weeks)

        // Carb target (race week forces carb-load; §4.5's race-type-specific
        // load values are deferred, so the flat 11 g/kg race-week load stands,
        // bounded below by FIX B).
        var carbsG = phase == .raceWeek
            ? 11.0 * kg
            : carbsPerKg(type, band: band) * kg

        // Protein: 1.7 g/kg baseline; recovery day 2.0 g/kg (§5.1 in-scope subset).
        // Order per §5.1: base × age modifier (§2.3) × diet digestibility modifier
        // (§2.4), then the 2.5 g/kg hard cap — all on the per-kg value, before the
        // fixed absolute activity additions below. Taper 1.85 / post-race 2.2 and
        // the dual-goal floor remain deferred.
        let baseProteinPerKg: Double = (type == "recovery_day" ? 2.0 : 1.7)
        let ageModifier = ageProteinModifier(user.age)
        let dietModifier = DietLayer.proteinModifier(for: user.diet)
        let proteinPerKg = min(baseProteinPerKg * ageModifier * dietModifier, 2.5)
        var proteinG = proteinPerKg * kg

        // Secondary activity adjustments (interim additive model — Section 11's
        // MET/EEE→TDEE routing is deferred; these deltas are applied pre-fat so
        // the Fat Engine reconciles the day's total to TDEE).
        for activity in additionalActivities {
            switch activity {
            case .gymUpper:             proteinG += 10
            case .gymLower:             proteinG += 15;  carbsG += 30
            case .gymFull:              proteinG += 15;  carbsG += 20
            case .other:                proteinG += 10;  carbsG += 15
            case .cycling, .swimming:   break   // already handled via canonical() if primary
            }
        }

        // Fat Engine (§6.1) — 20% floor / 35% ceiling + both reconciliation fixes.
        // isCarbLoad gates the training-day carb ceiling off during race-week
        // carb-load, whose deliberate TDEE overshoot is FIX B's job (see E1 note).
        let (fatG, adjustedCarbsG, fatFlags) = calculateFat(
            tdee: tdee, carbsG: carbsG, proteinG: proteinG, kg: kg, type: type,
            isCarbLoad: phase == .raceWeek
        )
        carbsG = adjustedCarbsG

        // Low-carb/keto conflict (§9.2): raise a coaching flag, never override the
        // carb target. Not reachable from onboarding (only omni/veg/vegan/pesc are
        // offered) but honoured for imported/edited profiles.
        var flags = fatFlags
        if DietLayer.isLowCarbConflict(user.diet) {
            flags.append(DietLayer.dietCarbConflictFlag)
        }

        // Day-before long-run/race carb boost (§4.2). Extra glycogen-topping carbs
        // for tomorrow's big session, surfaced on dinner by the Portion Engine
        // (Layer 8). Kept OUT of the reconciled daily totals above (per §4.4 it
        // rides separately) so the Fat Engine's FIX A/B can't claw it back.
        let dayBeforeBoostG = isDayBeforeLongOrRace(nextSessionType)
            ? dayBeforeBoostPerKg(band) * kg
            : 0.0

        return MacroTarget(
            carbsG: carbsG,
            proteinG: proteinG,
            fatG: fatG,
            caloriesKcal: carbsG * 4 + proteinG * 4 + fatG * 9,
            sessionType: resolvedType,
            trainingPhase: phase.rawValue,
            flags: flags,
            dayBeforeCarbBoostG: dayBeforeBoostG
        )
    }

    // Recovery-day detection (§3.4). Upgrades a *light* day that follows a long
    // run or race — deliberately does NOT downgrade a genuine hard session the
    // day after a long run (that would under-fuel it), which is a small,
    // intent-preserving deviation from the spec's literal unconditional override.
    static func resolveSessionType(today: String, previous: String?) -> String {
        guard let prev = previous,
              prev == "long_run" || prev == "race" || prev.hasPrefix("race_") else {
            return today
        }
        let hardTypes: Set<String> = [
            "tempo", "interval", "long_run",
            "race", "race_5k", "race_10k", "race_half", "race_marathon", "race_ultra"
        ]
        return hardTypes.contains(today) ? today : "recovery_day"
    }

    // True for session types that drive the primary TDEE/carb calc (runs + races,
    // incl. the split race_* values and the legacy "race"). Used by callers to
    // separate the primary run from logged activities.
    static func isRunSession(_ sessionType: String) -> Bool {
        switch sessionType {
        case "easy_run", "tempo", "interval", "long_run": return true
        case "race": return true
        default: return sessionType.hasPrefix("race_")
        }
    }

    static func isRaceSession(_ sessionType: String) -> Bool {
        sessionType == "race" || sessionType.hasPrefix("race_")
    }

    static func trainingPhase(weeksToRace: Int) -> TrainingPhase {
        switch weeksToRace {
        case let w where w > 12: return .baseBuilding
        case 8...12:             return .build
        case 4..<8:              return .peakTraining
        case 1..<4:              return .taper
        default:                 return .raceWeek
        }
    }

    static func weeksUntil(dateString: String) -> Int {
        let fmt = DateFormatter()
        fmt.dateFormat = "yyyy-MM-dd"
        fmt.timeZone = .current
        guard let raceDate = fmt.date(from: dateString) else { return 20 }
        let today = Calendar.current.startOfDay(for: Date())
        let days = Calendar.current.dateComponents([.day], from: today, to: raceDate).day ?? 0
        return max(0, days / 7)
    }

    // MARK: - Private

    // Maps aliases onto the canonical session types used by the lookup tables.
    // Legacy/ultra "race" strings fold to race_marathon so they can never fall
    // through to the rest default (defensive alias, master-spec flag 2).
    private static func canonical(_ sessionType: String) -> String {
        switch sessionType {
        case "cycling", "swimming":                     return "easy_run"
        case "race", "race_ultra", "race_ultra_marathon": return "race_marathon"
        default:                                        return sessionType
        }
    }

    // TDEE multipliers (§3.2). Existing run values unchanged; recovery_day + the
    // race_* split are the in-scope additions. The full §3.2 recalibration of
    // easy/tempo/interval/long is deferred.
    private static func tdeeMultiplier(_ type: String) -> Double {
        switch type {
        case "easy_run":       return 1.55
        case "tempo":          return 1.725
        case "interval":       return 1.725
        case "long_run":       return 1.9
        case "recovery_day":   return 1.35
        case "race_5k":        return 1.75
        case "race_10k":       return 1.85
        case "race_half":      return 2.00
        case "race_marathon":  return 2.40
        default:               return 1.2     // rest
        }
    }

    // Training-level → band (§2.1): beginner A / intermediate B / advanced C.
    // Any non-standard string falls to B, preserving the prior Band-B behavior.
    private static func trainingBand(_ level: String) -> String {
        switch level {
        case "beginner":  return "A"
        case "advanced":  return "C"
        default:          return "B"   // intermediate + unknown
        }
    }

    // Carbohydrate g/kg (§4.1) — full session × training-level band matrix. Only
    // the canonical primary session types are reachable here (cycling/swimming
    // fold to easy_run; gym_* are additive activities), so the §4.1 gym/cross-
    // training rows are intentionally omitted. Unknown band defaults to B.
    private static func carbsPerKg(_ type: String, band: String) -> Double {
        let (a, b, c): (Double, Double, Double)
        switch type {
        case "easy_run":       (a, b, c) = (4.0, 5.5, 6.5)
        case "tempo":          (a, b, c) = (5.5, 7.0, 8.0)
        case "interval":       (a, b, c) = (6.0, 7.5, 8.5)
        case "long_run":       (a, b, c) = (7.0, 8.5, 10.0)
        case "recovery_day":   (a, b, c) = (5.5, 6.0, 6.5)
        case "race_5k":        (a, b, c) = (5.0, 5.5, 6.0)
        case "race_10k":       (a, b, c) = (6.0, 6.5, 7.0)
        case "race_half":      (a, b, c) = (7.5, 8.5, 9.5)
        case "race_marathon":  (a, b, c) = (9.0, 10.0, 11.0)
        default:               (a, b, c) = (3.0, 3.5, 4.5)   // rest
        }
        switch band {
        case "A":  return a
        case "C":  return c
        default:   return b
        }
    }

    // Masters age → protein modifier (§2.3): <40 ×1.00 / 40–50 ×1.12 / >50 ×1.20.
    private static func ageProteinModifier(_ age: Int) -> Double {
        switch age {
        case ..<40:  return 1.00
        case 40...50: return 1.12
        default:     return 1.20   // > 50
        }
    }

    // Day-before boost trigger (§4.2): tomorrow is a long run or a half/full/ultra
    // race. Canonicalise first so legacy "race"/"race_ultra" fold to race_marathon;
    // short races (5k/10k) deliberately do NOT trigger a night-before load.
    private static func isDayBeforeLongOrRace(_ nextSessionType: String?) -> Bool {
        guard let next = nextSessionType else { return false }
        switch canonical(next) {
        case "long_run", "race_half", "race_marathon": return true
        default:                                       return false
        }
    }

    // Day-before boost g/kg by band (§4.2): A +1.0 / B +1.5 / C +2.0.
    private static func dayBeforeBoostPerKg(_ band: String) -> Double {
        switch band {
        case "A":  return 1.0
        case "C":  return 2.0
        default:   return 1.5
        }
    }

    // Session carb floors (§6.3) — the lower bound both fat guards reduce toward.
    private static func sessionCarbFloor(_ type: String) -> Double {
        switch type {
        case "long_run":                    return 5.0
        case "race_half":                   return 6.0
        case "race_marathon", "race_ultra": return 8.0
        default:                            return 3.0
        }
    }

    // Fat Engine (§6.1): derive fat from the 20–35%-of-energy band, then run
    // FIX A (ordinary fat-floor reconciliation) and FIX B (carb-load collision
    // guard) so the day's macros reconcile to TDEE. Returns (fat_g, adjusted
    // carbs_g, flags).
    private static func calculateFat(
        tdee: Double, carbsG: Double, proteinG: Double, kg: Double, type: String,
        isCarbLoad: Bool = false
    ) -> (fatG: Double, carbsG: Double, flags: [String]) {

        let usedCalories = carbsG * 4 + proteinG * 4
        let fatFromRemainder = max(0, (tdee - usedCalories) / 9)

        let fatMinimum = (tdee * 0.20) / 9   // 20% of energy (not 0.5 g/kg — §6.2)
        let fatMaximum = (tdee * 0.35) / 9

        var fatG = max(fatMinimum, min(fatFromRemainder, fatMaximum))
        var finalCarbsG = carbsG
        var flags: [String] = []

        // FIX A — when the fat floor triggers, reduce carbs toward the session
        // floor, then re-derive fat from the room actually left. Every session.
        if fatG > fatFromRemainder {
            flags.append("fat_floor_triggered")
            let excessFatCalories = (fatG - fatFromRemainder) * 9
            let carbReduction = excessFatCalories / 4
            let sessionFloorG = sessionCarbFloor(type) * kg
            finalCarbsG = max(carbsG - carbReduction, sessionFloorG)

            let remainingForFat = tdee - (finalCarbsG * 4 + proteinG * 4)
            fatG = max(remainingForFat / 9, fatMinimum * 0.9)
        }

        // FIX B — carb-load collision guard, TDEE-relative floor (not a flat
        // 8 g/kg, which overshoots small/low-TDEE bodies).
        let totalCheck = finalCarbsG * 4 + proteinG * 4 + fatMinimum * 9
        if totalCheck > tdee * 1.02 {
            let excess = totalCheck - tdee
            let reduction = excess / 4
            let weightBasedFloor = 8.0 * kg
            let tdeeRelativeFloor = (tdee * 0.55) / 4
            let hardFloor = min(weightBasedFloor, tdeeRelativeFloor)
            let newCarbs = max(finalCarbsG - reduction, hardFloor)
            if abs(newCarbs - finalCarbsG) > 0.0001 {
                flags.append("carb_load_capped")
                finalCarbsG = newCarbs
                let remainingForFat = tdee - (finalCarbsG * 4 + proteinG * 4)
                fatG = max(remainingForFat / 9, fatMinimum * 0.9)
            }
        }

        // Training-day carb ceiling (E1 monotonicity guard). On non-carb-load days,
        // never let reconciled carbs sit above the value that reconciles the day to
        // TDEE at the 20% fat floor. FIX A's 0.9×floor relaxation and FIX B's 2%
        // trigger tolerance can otherwise leave a lower training-level band above the
        // ceiling while a higher band gets clamped to it — inverting the §4.1 bands
        // (e.g. an 85kg long run, where advanced fell below intermediate). This cap
        // forbids the inversion; it is deliberately skipped on race-week carb-load,
        // whose intended TDEE overshoot is governed by FIX B. Full de-collapse of the
        // bands above the ceiling remains Section 7's job (AGENTS.md tech-debt E1).
        if !isCarbLoad {
            let carbCeiling = (tdee - proteinG * 4 - fatMinimum * 9) / 4
            let ceilingCap = max(carbCeiling, sessionCarbFloor(type) * kg)
            if finalCarbsG > ceilingCap + 0.0001 {
                flags.append("training_carb_ceiling_capped")
                finalCarbsG = ceilingCap
                let remainingForFat = tdee - (finalCarbsG * 4 + proteinG * 4)
                fatG = max(remainingForFat / 9, fatMinimum * 0.9)
            }
        }

        return (fatG, finalCarbsG, flags)
    }
}
